overlord/overlord.go

```diff
@@ -0,0 +1,108 @@
+// Copyright 2022 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package overlord
+
+import (
+	"context"
+
+	"github.com/juju/juju/overlord/logstate"
+	"github.com/juju/juju/overlord/schema"
+	"github.com/juju/juju/overlord/schema/updates"
+	"gopkg.in/tomb.v2"
+)
+
+// Overlord is the central manager of the system, keeping track
+// of all available state managers and related helpers.
+type Overlord struct {
+	stateEng *StateEngine
+	tomb     *tomb.Tomb
+	// managers
+	started   bool
+	schemaMgr *schema.SchemaManager
+}
+
+func newOverlord(s State, sch *schema.Schema) *Overlord {
+	o := &Overlord{
+		tomb:     new(tomb.Tomb),
+		stateEng: NewStateEngine(s),
+	}
+
+	// Ensure we register the new schema manager first.
+	o.schemaMgr = schema.NewManager(s, sch)
+	o.stateEng.AddManager(o.schemaMgr)
+
+	return o
+}
+
+// StartUp proceeds to run any expensive Overlord or managers initialization.
+// After this is done once it is a noop.
+func (o *Overlord) StartUp(ctx context.Context) error {
+	if o.started {
```

Comment:
You probably need a mutex here just in case `StartUp` gets called concurrently.

Reply:
Yeah, there is a potential race condition here.